### PR TITLE
add weighted selector.

### PR DESCRIPTION
**Pull request by qiwzhang**

### CODE REVIEW

File: contrib/endpoints/src/api_manager/weighted_selector.cc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  counts_.resize(list_.size());
+  std::fill(counts_.begin(), counts_.end(), 0);
+}
+
+float WeightedSelector::score(int index) {
+  return float(counts_[index]) / list_[index].second;
+}
+
+const std::string& WeightedSelector::Select() {
+  if (list_.size() == 1) {
```
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: What if the list were empty? There is no checks to ensure that an empty list is no a valid input.

File: contrib/endpoints/src/api_manager/weighted_selector.cc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    return list_[0].first;
+  }
+
+  auto min_score = score(0);
```
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: This assumes list cannot be empty.

File: contrib/endpoints/src/api_manager/weighted_selector.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  std::fill(counts_.begin(), counts_.end(), 0);
+}
+
+float WeightedSelector::score(int index) {
+  return float(counts_[index]) / list_[index].second;
+}
+
+const std::string& WeightedSelector::Select() {
+  if (list_.size() == 1) {
+    return list_[0].first;
```
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: Do you not have to increment counts in this case?

File: contrib/endpoints/src/api_manager/weighted_selector.cc (hunk 1 above)
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: I don't think you need the == 1 check. The code should work even if you don't have the check. Rather you should check for an empty list instead here.

Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: Added check with size=0 case,  special handling of size=1 is for optimization.

Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: But then you are not incrementing counts when size is 1. Is this intentional?

Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: Added check with size=0 case,  special handling of size=1 is for optimization

File: contrib/endpoints/src/api_manager/weighted_selector.cc (hunk 2 above)
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/weighted_selector.cc]: size=0 case should be handled before here.

**Comment by qiwzhang** [comment]
Yes, we don't need to update the count for size=1. it is not used.

**Pull request merged**